### DISCUSSION TITLE
Duda con parcial Copia.me

### DISCUSSION THREAD

**Discussion by ParedesMicaela** [dds-jv/foro#436] [Consultas de objetos] [answered]
<img width="884" alt="image" src="https://github.com/dds-jv/foro/assets/83614884/37d65ca2-2eda-4d0a-911f-6538df282f47">

Hola buenas! Tenia una duda con la resolucion de este parcial. Los distintos algoritmos de análisis son partes de mi sistema o son externos? Yo asumí que eran externos porque me parecía raro que el sistema tenga el código de estos algoritmos.  Suponiendo que son externos, no me cierra mucho que sean los tipos de documentos los que se analizan automáticamente porque eso significaría que cada  documento tiene que conocer a la biblioteca externa. Igualmente si son internos, tampoco me parece muy cohesivo que cada documento sepa todo el algoritmo que se usa para analizarse. No podría hacer por ejemplo una interfaz que engloble la forma automática de analizarse? O sea yo podria hacer algo de este estilo?
<img width="537" alt="image" src="https://github.com/dds-jv/foro/assets/83614884/ae3bd6e5-831a-4291-b294-4e984546fdfc">

**Comment by flbulgarelli** [comment, marked as answer]
:wave: ¡Hola!

> Los distintos algoritmos de análisis son partes de mi sistema o son externos?

Son parte del sistema. Eso no significa que tengas que implementarlos todos ellos durante el parcial, pero sí tenés que plantear como vas a hacer para soportarlos (de forma intercambiable), teniendo en cuenta la consideración de que son _costosos_ en términos de cómputo, es decir, requieren de una cantidad de ciclos de CPU no despreciable. 

> Yo asumí que eran externos porque me parecía raro que el sistema tenga el código de estos algoritmos. 

Es cierto que, por diferentes motivos, podrías decidir extraerlos a _nodos_ independientes en una _arquitectura distribuida_ (tema que retomaremos durante el segundo cuatrimestre), pero más allá de ello, no hay ningún impedimento para que sean parte del sistema: son algoritmos conocidos (sin ir mas lejos, por ejemplo la [biblioteca apache commons contiene una implementación de la distancia de Levenshtein](https://commons.apache.org/proper/commons-text/apidocs/org/apache/commons/text/similarity/LevenshteinDistance.html), la cual podrías ejecutar dentro de tu ambiente de objetos), no requieren interactuar con datos externos, ni están sujetos a licencias de servicio. 

En otras palabras, no debería resultar raro que estos algoritmos estén parcial o totalmente implementados, en mayor o menor medida, dentro de nuestro sistema, sobre todo si estos fueran elementos diferenciales de nuestro producto. Más aún: no veo buenos motivos para asumir que se ejecutan externamente. 

En todo caso, y releyendo tu pregunta, calculo que estás pensando en que los mismos estén implementado utilizando _bibliotecas_ externas. Eso sí podría ser posible (nuevamente, como en el caso de la distancia de Levenshtein que provee Apache Commons), pero no cambia sustancialmente a tu diseño. 

>  Suponiendo que son externos, no me cierra mucho que sean los tipos de documentos los que se analizan automáticamente porque eso significaría que cada documento tiene que conocer a la biblioteca externa. Igualmente si son internos, tampoco me parece muy cohesivo que cada documento sepa todo el algoritmo que se usa para analizarse.

Acá hay bastante tela para cortar. A priori, no veo un problema de cohesión, siempre y cuando la implementación del algoritmo sea sencilla o se delegue en una biblioteca (externa o no) que provea abstracciones stateless o fáciles de instanciar. Por ejemplo: 

```js
class Texto {
   validarAutomaticamente(otroDocumento) {
       return new LevenshteinDistance().apply(this.contenido, otroDocumento.contenido) < 0.5
   }  
}
```

Ahora bien, si la implementación fuera más compleja o difícil de instanciar, entonces sí entendería tu argumento en favor de la cohesión y valdría la pena extraer componentes independientes para realizar las validaciones. No estoy seguro de qué ventaja te daría que todos estos componentes implementaran una misma interfaz; de hecho, si para cada tipo de documento tuvieras exactamente un algoritmo validador, entonces incurrirías en un problema de _jerarquía paralelas_. Es más, ni siquiera se si podrías llevarlos todos a una misma interfaz fácilmente, dado que operarían con diferentes tipos de datos (secuencias de bytes, strings, etc) 

Por otro lado, si para cada tipo de documento existieran múltiples formas de validarlo, entonces ahí se justificaría desacoplar las ideas del documento y sus algoritmos de validación, para ganar flexibilidad. Pero ya estaríamos hablando de otro requerimiento y otro problema.

**Comment by flbulgarelli** [comment]
> No podría hacer por ejemplo una interfaz que engloble la forma automática de analizarse? O sea yo podria hacer algo de este estilo? <img alt="image" width="537" src="https://private-user-images.githubusercontent.com/83614884/342523666-ae3bd6e5-831a-4291-b294-4e984546fdfc.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTkzNDk3OTIsIm5iZiI6MTcxOTM0OTQ5MiwicGF0aCI6Ii84MzYxNDg4NC8zNDI1MjM2NjYtYWUzYmQ2ZTUtODMxYS00MjkxLWIyOTQtNGU5ODQ1NDZmZGZjLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA2MjUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNjI1VDIxMDQ1MlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTZlMzQxNTU5OTIwODliZjM5YWJkODY5NjY4YjQwMDQ1ODJlY2UyZDBhZmFkZjA3YTYwY2YyMjkyMjYwYWUxMmQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.59AsKLMoAqn44nZdIcsG7ork2L2Ns2pn8H-OIYB1pJM">

Acá, ojo, porque en el `haceLoTuyo`  estás haciendo _una trampa_ que se vuelve contra vos: el contrato es excesivamente amplio, un cheque en blanco. No está claro qué hace exactamente, qué parámetros recibe, de qué tipo, qué retorna. El nombre (que asumo, por contexto, que se refiere a un _Algoritmo**Automatico**DeValidacion_) no dice demasiado. En el pseudocódigo tampoco indicás qué hacés con su resultado (¿qué patrón de comunicación utiliza? ¿modifica el estado interno del algoritmo, lanza una excepción, devuelve un booleano o un objeto?) ni qué le pasás como argumento, lo que hace difícil entender la idea. 

De todas formas, vuelvo a lo que te comentaba anteriormente: no veo la ventaja de, en este punto, encapsular al algoritmo como un ente independiente del documento que valida bajo una interfaz común, ni tampoco me queda claro como harás para resolver la diversidad de tipos de datos y expresarlo todo bajo un tipo común. 

Decime si queda claro. 

¡Saludos!

**Comment by ParedesMicaela** [comment]
Perfecto quedó clarísimo, me la estaba complicando yo sola con los algoritmos 😅​😅​. Muchas gracias!
